signature/src/encoding.rs

```diff
@@ -0,0 +1,38 @@
+//! Encoding support.
+
+use crate::{Error, Result};
+
+#[cfg(feature = "alloc")]
+use alloc::{boxed::Box, vec::Vec};
+
+/// Support for decoding/encoding signatures as bytes.
+pub trait SignatureEncoding:
```

Comment:
This is largely a reaction (perhaps an overreaction) to mandatory trait bounds around the `Signature` trait in the v1.0 `signature` crate being too restrictive and preventing some things that would otherwise be valid. Although debatably the `AsRef<[u8]>` one was too onerous.

As far as a real-world example of something eliminating the bounds enables, the main thing is probably a heapless core implementation with an `alloc`-dependent `SignatureEncoding` impl:

https://github.com/RustCrypto/signatures/pull/562/files#diff-23f3fe37b15706957a041307578d2433204986a9eb42dc8d7c4126dcaf9266f1R226-R239

In this case, it's a variable-width ASN.1 DER-encoded type, which knows how to serialize itself into a buffer on heapless platforms, but when the heap is available, it's nice to be able to serialize it into a heap-backed structure (in this case `Box<[u8]>`)

Reply:
I see. In the above case, couldn't `SignatureEncoding` be implemented without the feature gate? The def simply won't contain the `to_vec` part if `alloc` isn't set. Or is the trait simply not useful in that case (for this particular crate)?

Reply:
The `SignatureEncoding::Repr` in that example is an `alloc::boxed::Box<[u8]>`, so no, it cannot be defined without the `alloc` feature enabled.